src/core/Sample.cpp

```diff
@@ -0,0 +1,488 @@
+	, m_startFrame(0)
+	, m_endFrame(m_buffer->size())
+	, m_loopStartFrame(0)
+	, m_loopEndFrame(m_buffer->size())
+{
+}
+
+Sample::Sample(const QByteArray& base64, int sampleRate)
+	: m_buffer(std::make_shared<SampleBuffer>(base64, sampleRate))
+	, m_startFrame(0)
+	, m_endFrame(m_buffer->size())
+	, m_loopStartFrame(0)
+	, m_loopEndFrame(m_buffer->size())
+{
+}
+
+Sample::Sample(const sampleFrame* data, int numFrames, int sampleRate)
+	: m_buffer(std::make_shared<SampleBuffer>(data, numFrames, sampleRate))
+	, m_startFrame(0)
+	, m_endFrame(m_buffer->size())
+	, m_loopStartFrame(0)
+	, m_loopEndFrame(m_buffer->size())
+{
+}
+
+Sample::Sample(std::shared_ptr<const SampleBuffer> buffer)
+	: m_buffer(buffer)
+	, m_startFrame(0)
+	, m_endFrame(m_buffer->size())
+	, m_loopStartFrame(0)
+	, m_loopEndFrame(m_buffer->size())
+{
+}
+
+Sample::Sample(const Sample& other)
+{
+	auto lock = std::shared_lock{other.m_mutex};
+	m_buffer = other.m_buffer;
+	m_startFrame = other.m_startFrame;
+	m_endFrame = other.m_endFrame;
+	m_loopStartFrame = other.m_loopStartFrame;
+	m_loopEndFrame = other.m_loopEndFrame;
+	m_amplification = other.m_amplification;
+	m_frequency = other.m_frequency;
+	m_reversed = other.m_reversed;
+}
+
+Sample::Sample(Sample&& other) noexcept
+{
+	auto lock = std::unique_lock{other.m_mutex};
+	m_buffer = std::move(other.m_buffer);
+	m_startFrame = std::exchange(other.m_startFrame, 0);
+	m_endFrame = std::exchange(other.m_endFrame, 0);
+	m_loopStartFrame = std::exchange(other.m_loopStartFrame, 0);
+	m_loopEndFrame = std::exchange(other.m_loopEndFrame, 0);
+	m_amplification = std::exchange(other.m_amplification, 0);
+	m_frequency = std::exchange(other.m_frequency, DefaultBaseFreq);
+	m_reversed = std::exchange(other.m_reversed, false);
+}
+
+Sample& Sample::operator=(Sample other) noexcept
+{
+	swap(*this, other);
+	return *this;
+}
+
+auto swap(Sample& first, Sample& second) -> void
+{
+	auto lock = std::scoped_lock{first.m_mutex, second.m_mutex};
+	using std::swap;
+	swap(first.m_buffer, second.m_buffer);
+	swap(first.m_startFrame, second.m_startFrame);
+	swap(first.m_endFrame, second.m_endFrame);
+	swap(first.m_loopStartFrame, second.m_loopStartFrame);
+	swap(first.m_loopEndFrame, second.m_loopEndFrame);
+	swap(first.m_amplification, second.m_amplification);
+	swap(first.m_frequency, second.m_frequency);
+	swap(first.m_reversed, second.m_reversed);
+}
+
+bool Sample::play(sampleFrame* dst, PlaybackState* state, int numFrames, float desiredFrequency, Loop loopMode) const
+{
+	auto lock = std::shared_lock{m_mutex};
+	if (numFrames <= 0 || desiredFrequency <= 0 || m_buffer->size() <= 0 || m_buffer->sampleRate() <= 0)
+	{
+		return false;
+	}
+
+	auto resampleRatio = static_cast<float>(Engine::audioEngine()->processingSampleRate()) / m_buffer->sampleRate();
+	resampleRatio *= m_frequency / desiredFrequency;
+
+	// If there happens to be an upper limit on the frames per period, we could use ArrayVector here instead
```

Comment:
> I think DEFAULT_BUFFER_SIZE might be the upper limit.

Yeah, that's right. I don't know if the upper limit should be 256 (I would think like 4096 or something), but that's that what code currently does right now.

Reply:
It seems that I was a bit wrong here. The size of ``playBuffer`` depends on the value of ``resampleRatio``, so I may not be able to use ``ArrayVector`` here without guessing what an optimal maximum size for it would be.